ssl/ssl_sess.c

```diff
@@ -44,6 +44,15 @@ static void SSL_SESSION_list_remove(SSL_CTX *ctx, SSL_SESSION *s);
 static void SSL_SESSION_list_add(SSL_CTX *ctx, SSL_SESSION *s);
 static int remove_session_lock(SSL_CTX *ctx, SSL_SESSION *c, int lck);
 
+/*
+ * TODO(TLS1.3): SSL_get_session() and SSL_get1_session() are problematic in
+ * TLS1.3 because, unlike in earlier protocol versions, the session ticket
+ * may not have been sent yet even though a handshake has finished. The session
+ * ticket data could come in sometime later...or even change if multiple session
+ * ticket messages are sent from the server. We need to work out how to deal
+ * with this.
```

Comment:
I think that David is describing behaviour similar to Postfix, which uses an external session cache via SSL_CTX_sess_set_new_cb():

https://github.com/vdukhovni/postfix/blob/master/postfix/src/tls/tls_client.c#L210

Mind you, I think that some applications do try to pull the session off the SSL handle directly.  I seem to recall some issues around changes in SSL_dup() as a result that had to be addressed.